src/feature/hs/hs_cell.c

```diff
@@ -802,6 +803,47 @@ get_introduce2_keys_and_verify_mac(hs_cell_introduce2_data_t *data,
   return intro_keys;
 }
 
+/** Return the newly allocated intro keys using the given service
+ * configuration and INTRODUCE2 data for the cell encrypted section.
+ *
+ * Every onion balance configured master key will be tried. If NULL is
+ * returned, no hit was found for the onion balance keys. */
+static hs_ntor_intro_cell_keys_t *
+get_intro2_keys_as_ob(const hs_service_config_t *config,
+                      const hs_cell_introduce2_data_t *data,
+                      const uint8_t *encrypted_section,
+                      size_t encrypted_section_len)
+{
+  uint8_t *ob_subcreds = NULL;
+  size_t ob_num_subcreds;
+  hs_ntor_intro_cell_keys_t *intro_keys = NULL;
+
+  ob_num_subcreds = hs_ob_get_subcredentials(config, &ob_subcreds);
```

Comment:
This is a design comment and perhaps a premature optimization, but given that we worry about performance, memory allocation and whatnot I'm gonna mention it. It seems non-optimal that we calculate all the subcredentials (which involves key blinding) even if we don't need them in the end. A more optimized approach would be to create a subcredential and then try it out, then go to the next one, etc. until we find the right one.

Reply:
It seems to me that we'd even be better off pre-computing all our subcredentials periodically, and not re-doing it every time we're doing a handshake.